### PR TITLE
Payment Request Button - Do not set WC session cookie and HTML container on the product page if not enabled

### LINKED ISSUE #4084: Payment Request Button - do not set the session cookie and HTML container on the product page if not enabled
See pd3Vha-37-p2#comment-498

When the Payment Request button is enabled, but set not to be visible on the product page:
Screenshot 2022-04-04 at 17 53 20

The [session cookie](https://github.com/Automattic/woocommerce-payments/blob/2751a8380866ae101e944924f881cac0b8032d95/includes/class-wc-payments-payment-request-button-handler.php#L157) and [the html containers](https://github.com/Automattic/woocommerce-payments/blob/2751a8380866ae101e944924f881cac0b8032d95/includes/class-wc-payments-payment-request-button-handler.php#L74-L75) should not be set/rendered.

### PR DESCRIPTION
**Pull request by ricardo**
Fixes #4084 

TL;DR: The VIP platform uses Varnish HTTP cache with cookie exceptions, such as `wp_woocommerce_session_*`. [Src](https://docs.wpvip.com/technical-references/caching/cookies/#:~:text=A%20WooCommerce%20session%20cookie).

That prevents product pages from being cached, even though express checkouts might be disabled for product pages.

#### Changes proposed in this Pull Request

- Do not set session cookie if product page is not enabled under settings.
- Check if HTML wrapper and separator should be displayed, [similar to Stripe extension](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/a8e04bcb4fd30d2a69b92067e10b4ace13849f0d/includes/payment-methods/class-wc-stripe-payment-request.php#L809) - this was probably ignored when customizable button locations was introduced.

#### Testing instructions

With Google Pay configured in your browser:

1. Make sure your site uses HTTPS.
2. As a merchant, make sure "Express checkouts" is enabled under the WCPay settings.
3. Under **"Express checkouts > Apple Pay / Google Pay"**, click "Customize".
4. Uncheck "Product page" and click "Save changes".
5. As a guest, navigate to a product page and remove all cookies (you can use a browser extension such as [EditThisCookie](https://chrome.google.com/webstore/detail/editthiscookie/fngmhnnpilhplaeedifhccceomclgfbg?hl=pt-BR), or **Chrome DevTools > Application > Cookie**).
6. Refresh the product page, and ensure the `wp_woocommerce_session_*` cookie is not created.
7. As a merchant, enable "Apple Pay / Google Pay" on the product page.
8. As a guest, navigate to a product page and ensure the `wp_woocommerce_session_*` cookie was created.
9. Proceed with Google Pay checkout and make sure it works.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

File: includes/class-wc-payments-payment-request-button-handler.php [hunk 1]
```diff
@@ -150,7 +150,10 @@ public function get_total_label() {
 	 * @return void
 	 */
 	public function set_session() {
-		if ( ! $this->is_product() || ( isset( WC()->session ) && WC()->session->has_session() ) ) {
+		if (
+			! ( $this->is_product() && $this->is_available_at( 'product' ) )
```
Comment (vbelolapotkov) [inline includes/class-wc-payments-payment-request-button-handler.php:154]: @ricardo I've noticed there is another condition for product page in `should_show_payment_request_button` method [here](https://github.com/Automattic/woocommerce-payments/blob/d48df11bcc2c50ce931a699371390eb7ef7a2cac/includes/class-wc-payments-payment-request-button-handler.php#L431-L435), should we take that into account as well? So this line would look like this:

```
! ( $this->is_product() && $this->should_show_payment_request_button() )
```

IMO that would keep all the logic in the same place.

Comment (vbelolapotkov) [reply includes/class-wc-payments-payment-request-button-handler.php:154]: Another thing worth doing here is to add a comment explaining why there is such a special treatment for product page. Something like: `Don't set sessions cookies on product pages to allow caching when payment request buttons are disabled. But keep cookies if there is already active WC session in place.`

**Review by vbelolapotkov — COMMENTED** [review]
I haven't tested the changes, but the code looks good to me (see a couple of comments). I'll ask someone else to run tests.

File: includes/class-wc-payments-payment-request-button-handler.php (hunk 1 above)
Comment (ricardo) [reply includes/class-wc-payments-payment-request-button-handler.php]: Good call. Makes sense to use `should_show_payment_request_button`.

Made the change in [8ff6ba9](https://github.com/Automattic/woocommerce-payments/pull/4085/commits/8ff6ba97e2cc006684775da76599718bab2c7d3a).

**Review by vbelolapotkov — APPROVED** [review]
Approving since the code looks good to me, but I haven't tested. So I'd expect someone else to test it properly and approve before merging.

**Review by cesarcosta99 — APPROVED** [review]
LGTM and tests well! :shipit:
